File: 300. Longest Increasing Subsequence.py
#Python:

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp = [1] * len(nums)
        for i in range(1, len(nums)):
            for j in range(0, i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)
        res = 1
        for x in dp:
            res = max(res, x)
        return res

#Java:

class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        int[] dp = new int[n];
        for(int i=0; i<n; i++){
            for(int j=0; j<i; j++){
                if(nums[i] > nums[j]){
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
        }
        int res = 0;
        for(int num : dp){
            res = Math.max(res, num);
        }
        return res + 1;
    }
}

File: 300. Longest Increasing Subsequence.java
class Solution {
    int[] dp;
    public List<Integer> largestDivisibleSubset(int[] nums) {
        int ldsSize = getLdsSize(nums);
        return constructLds(nums, ldsSize);
    }
    private int getLdsSize(int[] nums){
        Arrays.sort(nums);
        dp = new int[nums.length];
        Arrays.fill(dp, 1);
        int lds = 1;
        for(int i=1; i<nums.length; i++){
            for(int j=0; j<i; j++){
                if(nums[i] % nums[j] == 0){
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
                lds = Math.max(dp[i], lds);
            }
        }
        return lds;
    }
    private List<Integer> constructLds(int[] nums, int ldsSize){
        var res = new LinkedList<Integer>();
        int prev = -1;
        for(int i=nums.length-1; i>=0; i--){
            if(dp[i] == ldsSize && (prev == -1 || prev % nums[i] == 0)){
                res.addFirst(nums[i]);
                prev = nums[i];
                ldsSize--;
            }
        }
        return res;
    }
}
